synapse/util/task_scheduler.py

```diff
@@ -0,0 +1,244 @@
+            )
+
+        launch_now = False
+        if timestamp is None or timestamp < self.clock.time_msec():
+            timestamp = self.clock.time_msec()
+            launch_now = True
+
+        task = ScheduledTask(
+            random_string(16),
+            action,
+            TaskStatus.SCHEDULED,
+            timestamp,
+            resource_id,
+            params,
+            None,
+            None,
+        )
+        await self.store.upsert_scheduled_task(task)
+
+        if launch_now and self.run_background_tasks:
+            await self._launch_task(task, True)
+
+        return task.id
+
+    async def update_task(
+        self,
+        id: str,
+        *,
+        status: Optional[TaskStatus] = None,
+        result: Optional[JsonMapping] = None,
+        error: Optional[str] = None,
+    ) -> None:
+        """Update some task associated values.
+
+        This is used internally, and also exposed publically so it can be used inside task functions.
+        This allows to store in DB the progress of a task so it can be resumed properly after a restart of synapse.
+
+        Args:
+            id: the id of the task to update
+            status: the new `TaskStatus` of the task
+            result: the new result of the task
+            error: the new error of the task
+        """
+        await self.store.update_scheduled_task(
+            id,
+            timestamp=self.clock.time_msec(),
+            status=status,
+            result=result,
+            error=error,
+        )
+
+    async def get_task(self, id: str) -> Optional[ScheduledTask]:
+        """Get a specific task description by id.
+
+        Args:
+            id: the id of the task to retrieve
+
+        Returns: the task description or `None` if it doesn't exist
+            or it has already been cleaned
+        """
+        return await self.store.get_scheduled_task(id)
+
+    async def get_tasks(
+        self, action: str, resource_id: Optional[str]
+    ) -> List[ScheduledTask]:
+        """Get a list of tasks associated with an action name, and
+        optionally with a resource id.
+
+        Args:
+            action: the action name of the tasks to retrieve
+            resource_id: if `None`, returns all associated tasks for
+                the specified action name, regardless of the resource id
+
+        Returns: a list of `ScheduledTask`
+        """
+        return await self.store.get_scheduled_tasks(action, resource_id)
+
+    async def _scheduled_tasks_loop(self) -> None:
+        """Main loop taking care of launching the scheduled tasks when needed."""
+        for task in await self.store.get_scheduled_tasks():
+            if task.id not in self.running_tasks:
+                if (
+                    task.status == TaskStatus.SCHEDULED
+                    and task.timestamp < self.clock.time_msec()
+                ):
+                    await self._launch_task(task, True)
```

Comment:
Booleans floating without a keyword argument are confusing:

```suggestion
                    await self._launch_task(task, first_launch=True)
```

(And same below/above)